### PR TITLE
[IOTDB-5559] Implement metric exporters for RatisConsensus

### PR DESCRIPTION
**Pull request by SzyWilliam**
See https://apache-iotdb.feishu.cn/docx/NHsMdWyaJoxjcFxICoCcvDannNb

### CODE REVIEW

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    } catch (IllegalArgumentException ignored) {
+    }
+    try {
+      metricService.remove(MetricType.TIMER, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
+    try {
+      metricService.remove(MetricType.AUTO_GAUGE, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
```
Comment (SpriCoder) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java:98]: If we ignore exceptions here, why use three try?

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/MetricRegistryManager.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.Consumer;
+
+public class MetricRegistryManager extends MetricRegistries {
+
+  private static final Logger LOG = LoggerFactory.getLogger(MetricRegistryManager.class);
```
Comment (SpriCoder) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/MetricRegistryManager.java:41]: Maybe we use LOGGER in our project?

**Review by SpriCoder — CHANGES_REQUESTED** [review]
PTAL~
Please add the description of metrics at the same time

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java (hunk 1 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java:97]: To make each of them independent

**Comment by SzyWilliam** [comment]
Thanks for the reviews @SpriCoder . Made changes on code, please take a look again. Also cc @OneSizeFitsQuorum @BUAAserein

**Review by SpriCoder — APPROVED** [review]
LGTM~

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    gaugeCache.clear();
+    timerCache.clear();
+  }
+
+  @Override
+  public Gauge gauge(String name, MetricRegistry.MetricSupplier<Gauge> metricSupplier) {
+    final String fullName = getMetricName(name);
+    final GaugeProxy gauge = new GaugeProxy(metricSupplier);
+    final GaugeProxy previous = gaugeCache.putIfAbsent(fullName, gauge);
+    if (previous == null) {
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java:117]: Please ensure atomicity.

**Review by OneSizeFitsQuorum — COMMENTED** [review]
PTAL

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/CounterProxy.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ * under the License.
+ */
+package org.apache.iotdb.consensus.ratis.metrics;
+
+import org.apache.iotdb.metrics.type.Counter;
+
+/** A Proxy class using IoTDB Counter to replace the dropwizard Counter. */
+public class CounterProxy extends com.codahale.metrics.Counter {
+
+  // IoTDB Counter
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/CounterProxy.java:26]: done

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/TimerProxy.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+public class TimerProxy extends com.codahale.metrics.Timer {
+  private final Timer timer;
+
+  TimerProxy(Timer timer) {
+    this.timer = timer;
+  }
+
+  @Override
+  public Context time() {
+    return super.time();
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/TimerProxy.java:34]: The `time` method is only used as a user time clock. Will add comments.

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/RatisMetricSet.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ * under the License.
+ */
+package org.apache.iotdb.consensus.ratis.metrics;
+
+import org.apache.iotdb.metrics.AbstractMetricService;
+import org.apache.iotdb.metrics.metricsets.IMetricSet;
+
+import org.apache.ratis.metrics.MetricRegistries;
+
+public class RatisMetricSet implements IMetricSet {
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/RatisMetricSet.java:26]: This class will be used in subsequent PRs.

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/MetricRegistryManager.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+import org.apache.ratis.util.TimeDuration;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+
+public class MetricRegistryManager extends MetricRegistries {
+  private final RefCountingMap<MetricRegistryInfo, RatisMetricRegistry> registries;
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/MetricRegistryManager.java:36]: Done

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    } catch (IllegalArgumentException ignored) {
+    }
+    try {
+      metricService.remove(MetricType.TIMER, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
+    try {
+      metricService.remove(MetricType.AUTO_GAUGE, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java:97]: Will add comments.

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  @Override
+  public JmxReporter getJmxReporter() {
+    throw new UnsupportedOperationException("JmxReporter is not used in Ratis Metrics");
+  }
+
+  @Override
+  public void setConsoleReporter(ConsoleReporter consoleReporter) {
+    // Not Implemented: ConsoleReporter is not used in Ratis Metrics
```
[the remark this answers is not in the document]
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java:192]: done

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java (hunk 3 above)
Comment (SzyWilliam) [reply consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java:117]: Thanks for pointing out this! Will switch to computeIfAbsent

**Comment by SzyWilliam** [comment]
@OneSizeFitsQuorum Thanks a lot for the careful reviews. Made changes on code.

**Review by BUAAserein — APPROVED** [review]
LGTM~

File: consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/TimerProxy.java [hunk 10]
```diff
@@ -30,6 +30,7 @@ public class TimerProxy extends com.codahale.metrics.Timer {
   }
 
   @Override
+  // time() method is used as a user time clock. Will reuse the dropwizard implementation.
```
Comment (OneSizeFitsQuorum) [inline consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/TimerProxy.java:33]: Pleause /** / rather than //

**Review by OneSizeFitsQuorum — APPROVED** [review]

**Pull request merged**